components/cloud/azure/src/azblob.rs

```diff
@@ -0,0 +1,740 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+use async_trait::async_trait;
+use azure_core::auth::{TokenCredential, TokenResponse};
+use azure_core::prelude::*;
+use azure_identity::token_credentials::{ClientSecretCredential, TokenCredentialOptions};
+use azure_storage::blob::prelude::*;
+use azure_storage::core::{prelude::*, ConnectionStringBuilder};
+use chrono::{Duration as chDuration, Utc};
+use cloud::blob::{
+    none_to_empty, BlobConfig, BlobStorage, BucketConf, PutResource, StringNonEmpty,
+};
+use futures_util::{
+    io::{AsyncRead, AsyncReadExt},
+    stream,
+    stream::StreamExt,
+    TryStreamExt,
+};
+pub use kvproto::brpb::{AzureBlobStorage as InputConfig, Bucket as InputBucket, CloudDynamic};
+use oauth2::{ClientId, ClientSecret};
+use tikv_util::debug;
+use tikv_util::stream::{retry, RetryError};
+use tokio::sync::Mutex;
+use tokio::time::{timeout, Duration};
+
+use std::str::FromStr;
+use std::{
+    env, io,
+    sync::{Arc, RwLock},
+};
+
+const ENV_CLIENT_ID: &str = "AZURE_CLIENT_ID";
+const ENV_TENANT_ID: &str = "AZURE_TENANT_ID";
+const ENV_CLIENT_SECRET: &str = "AZURE_CLIENT_SECRET";
+const ENV_ACCOUNT_NAME: &str = "AZURE_STORAGE_ACCOUNT";
+const ENV_SHARED_KEY: &str = "AZURE_STORAGE_KEY";
+
+/// CredentialInfo saves the credential variables from the enviroment:
+/// client_id:      $AZURE_CLIENT_ID,
+/// tenant_id:      $AZURE_TENANT_ID,
+/// client_secret:  $AZURE_CLIENT_SECRET,
+#[derive(Clone)]
+struct CredentialInfo {
+    client_id: ClientId,
+    tenant_id: String,
+    client_secret: ClientSecret,
+}
+
+impl std::fmt::Debug for CredentialInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CredentialInfo")
+            .field("client_id", &self.client_id)
+            .field("tenant_id", &self.tenant_id)
+            .field("client_secret", &"?")
```

Comment:
just use `#[derive(Debug)]`. 

`oauth2::ClientSecret` is a "secret type" and the debug representation is `"ClientSecret([redacted])"` already.